File: file_analyze.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func percentage(num, total int) float64 {
	return (float64(num) / float64(total)) * 100
}


func GroupHash(hashFile string, truncLine string, ignoreStr string, extractInt int) {
	var hashtype string
	var err error
	hash_counts := make(map[string]int)
	total_hashes := 0
	countLine := 0

	file, err := os.Open(hashFile)
	defer file.Close()
	if err != nil {
		fmt.Printf(ifColor(bred, "[!] Error: %s\n", rst), err)
		return
	}

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		input := strings.TrimSpace(scanner.Text())
		countLine++

		if input != "" {
			cleanLine := input

			if truncLine != "" || ignoreStr != ""{    
					cleanLine = CleanHashLine(cleanLine, truncLine, ignoreStr, countLine, "default")
					if len(cleanLine) > 0 {
					                  total_hashes++
					                //  fmt.Println("Test: ", cleanLine)
					 
					                  hashtype = DefaultFormat(cleanLine)
					 
					                  hash_counts[hashtype]++
				}
				} 
					

				if extractInt > 0 {
					extractedHashes := ExtractHexHashes(cleanLine, extractInt)

					for _, extractedHash := range extractedHashes {
						if len(extractedHash) > 0 {
							fmt.Println(ifColor(bgrn, "\n Extracted Hash: ", rst), ifColor(bylw, extractedHash, rst))
							if len(extractedHash) > 0 {
								total_hashes++

								hashtype = DefaultFormat(extractedHash)

								hash_counts[hashtype]++
							}
						}
					}
					continue

				}

			if len(cleanLine) > 0 {
				total_hashes++
			//	fmt.Println("Test: ", cleanLine)

				hashtype = DefaultFormat(cleanLine)

				hash_counts[hashtype]++
			}
		}
	}

	if total_hashes > 0 {
		fmt.Print(ifColor(bgrn, "\n[~] Found ", rst) + ifColor(borng, total_hashes, rst) + ifColor(bgrn, " hashes in ", rst) + ifColor(borng, hashFile, rst))
	}
	for hashtype, count := range hash_counts {
		pct := percentage(count, total_hashes)
		fmt.Printf(ifColor(bgrn, "\n\n{", rst)+ifColor(bylw, "%.2f%%", rst)+ifColor(bgrn, "} ", rst)+ifColor(bcyn, "%d/%d", rst)+ifColor(bcyn, " Of The Hashes Are:\n", rst)+"\n%s", pct, count, total_hashes, hashtype)
	}
	if err := scanner.Err(); err != nil {
		fmt.Printf(ifColor(bred, "[!] Error: %s\n", rst), err)
	}
}


func VerboseAnalyze(hashLine, truncLine, ignoreStr, format string) {
	var hashtype string

	if hashLine != "" {

		if truncLine != "" || ignoreStr != "" {

			if truncLine != "" || ignoreStr != "" {
				hashLine = CleanHashLine(hashLine, truncLine, ignoreStr, /*countLine*/ 0, format)
			}
		}

		if len(hashLine) > 0 {
			if format == "csv" {
				CSVFormat(hashLine)
			} else if format == "json" {
				hashtype = JSONFormat(hashLine)
				fmt.Println(hashtype)
			} else if format == "default" {
				fmt.Println(ifColor(bcyn, "\nPOSSIBLE HASH TYPES", rst))
				hashtype = DefaultFormat(hashLine)
				fmt.Println(hashtype)
			}
		}
	}
}

func FileVerboseAnalyze(hashFile, truncLine, ignoreStr, format string) {
	countLine := 0
	file, err := os.Open(hashFile)
	defer file.Close()
	if err != nil {
		fmt.Printf(ifColor(bred, "[!] Error: %s\n", rst), err)
		return
	}

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		hashLine := strings.TrimSpace(scanner.Text())
		countLine++
		if truncLine != "" || ignoreStr != "" {
			hashLine = CleanHashLine(hashLine, truncLine, ignoreStr, countLine, format)
		}
		if format == "default"{
		fmt.Print(ifColor(bgrn, "\nHash: ", rst), ifColor(bylw, hashLine, rst))
		}
		VerboseAnalyze(hashLine, "", "", format)
	}
	if err := scanner.Err(); err != nil {
		fmt.Printf(ifColor(bred, "[!] Error: %s\n", rst), err)
	}
}

File: src/libpeek/file_analyze.nim
import strformat, strutils, tables, os, terminal
import outfmt, analyze, ../cli_flag

proc percentage(num, total: int): float = 
    return float(num/total)*100.0


#[proc fileIdentifyGroupWithExtract*(hash_file, extCtext: string, extHex: int) = 
  var 
    hashtype: string
    hash_counts = initTable[string, int]()
    totalHashes = 0
    extValue: seq[string]

  if not hash_file.fileExists():
    stderr.write(ifColor(fgRed, "[ERROR] ") & ifColor(fgGreen, fmt "{hashfile} not found\n"))
    stderr.flushFile()
    return

  #should check if its a directory

  let file = open(hashfile, fmRead)
  defer: file.close()

  for line in file.lines:
    let input = line.strip()

    if input != "":
      totalHashes.inc()  
      if extCtext != "" or extHex > 0:
        extValue.add(myExtractor.extract(input))
        for extStr in extValue:
        #if format == "" or format == "default":
          stdout.write(ifColor(fgGreen, "\nExtracted Hash: ") & ifColor(fgYellow, fmt "{extStr}"))
          stdout.flushFile()
      hashtype = defaultFormat(identify(extStr))
      #[elif format == "json":
        echo jsonFormat(identify(input))
        return
      elif format == "csv":
        echo csvFormat(identify(input))]#
      #check existence of key before incrementing
      if hashtype in hash_counts:
        inc(hash_counts[hashtype])
      else:
        hash_counts[hashtype] = 1
  if total_hashes > 0:
    stdout.write(ifColor(fgCyan, "\n[INFO]") & ifColor(fgGreen, " Found ") & ifColor(fgCyan, totalHashes) & ifColor(fgGreen, " hashes in ") & ifColor(fgYellow, fmt "{hash_file}\n"))
    stdout.flushFile()
    for hashtype, count in hash_counts:
      let pct = percentage(count, totalHashes)
      stdout.write(ifColor(fgGreen, "\n{") & ifColor(fgYellow, fmt "{pct:.2f}%") & ifColor(fgGreen, "}") & ifColor(fgCyan, fmt " {count}/{totalHashes} Of The Hashes Are:\n") & fmt "{hashtype}\n")
      stdout.flushFile()

]#


proc fileIdentifyGroup*(hash_file: string) = 
  var 
    hashtype: string
    hash_counts = initTable[string, int]()
    totalHashes = 0
  if not hash_file.fileExists():
    stderr.write(ifColor(fgRed, "[ERROR] ") & ifColor(fgGreen, fmt "{hashfile} not found\n"))
    stderr.flushFile()
    return

  #should check if its a directory

  let file = open(hashfile, fmRead)
  defer: file.close()

  for line in file.lines:
    let input = line.strip()

    if input != "":
      totalHashes.inc()
      #if format == "" or format == "default":
      hashtype = defaultFormat(identify(input))
      #[elif format == "json":
        echo jsonFormat(identify(input))
        return
      elif format == "csv":
        echo csvFormat(identify(input))]#
      #check existence of key before incrementing
      if hashtype in hash_counts:
        inc(hash_counts[hashtype])
      else:
        hash_counts[hashtype] = 1
  if total_hashes > 0:
    stdout.write(ifColor(fgCyan, "\n[INFO]") & ifColor(fgGreen, " Found ") & ifColor(fgCyan, totalHashes) & ifColor(fgGreen, " hashes in ") & ifColor(fgYellow, fmt "{hash_file}\n"))
    stdout.flushFile()
    for hashtype, count in hash_counts:
      let pct = percentage(count, totalHashes)
      stdout.write(ifColor(fgGreen, "\n{") & ifColor(fgYellow, fmt "{pct:.2f}%") & ifColor(fgGreen, "}") & ifColor(fgCyan, fmt " {count}/{totalHashes} Of The Hashes Are:\n") & fmt "{hashtype}\n")
      stdout.flushFile()


proc fileIdentify*(hashFile, format: string) = 
  if not hashFile.fileExists():
    stderr.write(ifColor(fgRed, "[ERROR] ") & ifColor(fgGreen, fmt "{hashfile} not found\n"))
    stderr.flushFile()
    return

  let file = open(hashFile, fmRead)
  defer: file.close()

  for line in file.lines():
    if line.strip() != "":
      if format == "" or format == "default":
        stdout.write(fgCyan, "POSSIBLE HASHTYPES")
        stdout.flushFile()
        echo defaultFormat(identify(line.strip()))
      elif format == "json":
        echo jsonFormat(identify(line.strip()))
        return
      elif format == "csv":
        echo csvFormat(identify(line.strip()))
      